api/concept/Value.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (C) 2022 Vaticle
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.vaticle.typedb.client.api.concept;
+
+import com.vaticle.typedb.client.common.NativeObject;
+import com.vaticle.typedb.client.common.exception.TypeDBClientException;
+
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.ZoneOffset;
+
+import static com.vaticle.typedb.client.common.exception.ErrorMessage.Concept.BAD_ENUM_VALUE;
+import static com.vaticle.typedb.client.jni.typedb_client.value_equals;
+import static com.vaticle.typedb.client.jni.typedb_client.value_get_boolean;
+import static com.vaticle.typedb.client.jni.typedb_client.value_get_date_time_as_millis;
+import static com.vaticle.typedb.client.jni.typedb_client.value_get_double;
+import static com.vaticle.typedb.client.jni.typedb_client.value_get_long;
+import static com.vaticle.typedb.client.jni.typedb_client.value_get_string;
+import static com.vaticle.typedb.client.jni.typedb_client.value_is_boolean;
+import static com.vaticle.typedb.client.jni.typedb_client.value_is_date_time;
+import static com.vaticle.typedb.client.jni.typedb_client.value_is_double;
+import static com.vaticle.typedb.client.jni.typedb_client.value_is_long;
+import static com.vaticle.typedb.client.jni.typedb_client.value_is_string;
+import static com.vaticle.typedb.client.jni.typedb_client.value_new_boolean;
+import static com.vaticle.typedb.client.jni.typedb_client.value_new_date_time_from_millis;
+import static com.vaticle.typedb.client.jni.typedb_client.value_new_double;
+import static com.vaticle.typedb.client.jni.typedb_client.value_new_long;
+import static com.vaticle.typedb.client.jni.typedb_client.value_new_string;
+
+public class Value extends NativeObject<com.vaticle.typedb.client.jni.Value> {
+    private final int hash;
+
+    public Value(com.vaticle.typedb.client.jni.Value value) {
+        super(value);
+        this.hash = toString().hashCode();
```

Comment:
Let's use well-defined hash and equality for the `Concept` layer objects, since they should be pretty simple. They must be aligned: `object.equals(obj)` => `object.hashCode() == obj.hashCode()`